### PR TITLE
fix(compiler-cli): ensure the compiler tracks `ts.Program`s correctly

### PR DESCRIPTION
**Pull request by alxhub**
`NgCompiler` previously had a notion of the "next" `ts.Program`, which
served two purposes:

* it allowed a client using the `ts.createProgram` API to query for the
  latest program produced by the previous `NgCompiler`, as a starting
  point for building the _next_ program that incorporated any new user
  changes.

* it allowed the old `NgCompiler` to be queried for the `ts.Program` on
  which all prior state is based, which is needed to compute the delta
  from the new program to ultimately determine how much of the prior
  state can be reused.

This system contained a flaw: it relied on the `NgCompiler` knowing when
the `ts.Program` would be changed. This works fine for changes that
originate in `NgCompiler` APIs, but a client of the `TemplateTypeChecker`
may use that API in ways that create new `ts.Program`s without the
`NgCompiler`'s knowledge. This caused the `NgCompiler`'s concept of the
"next" program to get out of sync, causing incorrectness in future
incremental analysis.

This refactoring cleans up the compiler's `ts.Program` management in
several ways:

* `TypeCheckingProgramStrategy`, the API which controls `ts.Program`
  updating, is renamed to the `ProgramDriver` and extracted to a separate
  ngtsc package.

* It loses its responsibility of determining component shim filenames. That
  functionality now lives exclusively in the template type-checking package.

* The "next" `ts.Program` concept is renamed to the "current" program, as
  the "next" name was misleading in several ways.

* `NgCompiler` now wraps the `ProgramDriver` used in the
  `TemplateTypeChecker` to know when a new `ts.Program` is created,
  regardless of which API drove the creation, which actually fixes the bug.

### CODE REVIEW

File: packages/compiler-cli/src/ngtsc/program_driver/README.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+`ProgramDriver` is a small but important interface which allows the template type-checking machinery to request changes to the current `ts.Program`, and to receive a new `ts.Program` with those changes applied. This is used to add template type-checking code to the current `ts.Program`, eventually allowing for diagnostics to be produced within that code. This operation is abstracted behind this interface because different clients create `ts.Program`s differently. The Language Service, for example, creates `ts.Program`s from the current editor state on request, while the TS compiler API creates them explicitly.
+
+When running using the TS APIs, it's important that each new `ts.Program` be created incrementally from the previous `ts.Program`. Under a normal compilation, this means that programs alternate between template type checking programs and user programs:
+
+* `ts.Program#1` is created from user input (the user's source files).
+* `ts.Program#2` is created incrementally on top of #1 for template type-checking, and adds private TCB code.
+* `ts.Program#3` is created incrementally on top of #2 when the user makes changes to files on disk (incremental build).
```
Comment (kyliau) [inline packages/compiler-cli/src/ngtsc/program_driver/README.md:9]: If the input program (program#1) hasn't changed, but there are some operations that cause additional private TCB code to be generated **after** step 2, is the new program generated based on (1) or (2)?
Could you please clarify this in the example above?

Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/program_driver/README.md:9]: This example is describing "under a normal compilation", meaning the normal functioning of `ngc` with no usage of `TemplateTypeChecker` APIs directly.

Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/program_driver/README.md:9]: I added a section to clarify how such a client might consume `TsCreateProgramDriver` and manage programs that way.

**Review by kyliau — APPROVED** [review]

**Review by josephperrott — APPROVED** [review]
LGTM

Reviewed-for: dev-infra

**Pull request closed without merging**